proposals/closures.md

```diff
@@ -0,0 +1,256 @@
+        def g() -> int:
+            return h()
+        def h() -> int:
+            return x + 1
+        return g
+
+    f(10)()
+    ```
+
+## How we will add tests for your feature
+We will check the test cases’ final outputs, results of type checking and intermediate ASTs generated from escaping analysis. In addition, test cases for scenarios where syntax, type and runtime error are arised are covered.
+
+## New AST forms planing to add
+``` typescript
+export type Type = 
+    ...
+	| { tag: “callable”, args: Array<Type>, ret: Type }
+
+export type Scope<A> 
+  = { a?: A, tag: "global", name: string}
+  | { a?: A, tag: "nonlocal", name: string}
+
+export type FunDef<A> = {
+	a?: A,
+	name: string,
+	parameters: Array<Parameter<A>>,
+	ret: Type,
+  decls: Array<Scope<A>>,
+	inits: Array<VarInit<A>>,
+	funs: Array<FunDef<A>>,
+	body: Array<Stmt<A>>
+}
+
+export type VarInit<A> = 
+  | { a?: A, name: string, type: Type, value: Literal }
+
+export type Expr<A> = ...
+	// note that Python's type hint does not annotate lambda
+	// However, it does allow fancy calling (e.g. opt args)
+	// TODO: talk to fancy calling conventions group about this 
+	| { tag: "lambda", args: string[], ret: Expr<A> }
+
+export type Value = 
+    ...
+	| { tag: "callable", name: string, addr: number } // closure
+```
+
+## New functions, datatypes, and/or files added to the codebase
+
+- eair.ts
+  ``` typescript
+  // 'eair' stands for the intermedia representation after escaping analysis
+  
+  // Defines an AST that is similar to the input AST of ea (defined
+  // in the following file) but without callable type, nested function
+  // as well as nonlocal keywords, while other non-trivial changes 
+  // (some of which might be useful for optimization) is listed.
+  
+  // nonlocals will be used as the parameters of the constructor
+  // use the FunDef without nested function
+  export type Closure<A> = 
+	  { a?: A, name: string, nonlocals: Array<VarInit<A>>, apply: FunDef<A>> }
+  
+  // change case, TODO: design decisions to discuss
+  export type Expr<A> = ...
+	  | { a?: A, tag: "ref", name: string }
+	  | { a?: A, tag: "fun_id", name: string }
+  ```
+ 
+- ea.ts
+  ``` typescript
+  type EaInput = ... // typed ast or ir from another group
+  export function ea(ast: EaInput): eair.Program {
+	  // escape analysis here
+  }
+  ```
+
+## Changes to existing functions, datatypes, and/or files in the codebase
+We need to change the `parser.ts` to support recursively parsing nested fucntions;
+`tc.ts` for the new `callable` type and so on; `compiler.ts` to support the new features added.
+
+## Value representation and memory layout for any new runtime values
+The memory layout of closures and references should be as similar as that of class (reference wrapper and closure implemented as something like non-inheritable class). As for non-escaping functions using as values, they will be an i32 pointer pointing to the corresponding functions in the table.
```

Comment:
This will require a little bit of careful design, but not too much I hope. We can use the extra-argument infrastructure for method calls to pass an extra reference into the function that holds all the nonlocal variables.

The main thing is that we don't have vtable support yet, and waiting for the inheritance group isn't a good idea. So maybe just having the first word of the closure as the function pointer would work well. That would also keep their representation working the same as function values that _don't_ close over any nonlocal variables, so the indirect_call's index is always found by loading the first word of memory referenced by the function value (whether closure or regular function).

That sounds like a good place to start.

Reply:
Great. Regular callable variable is a pointer pointing to the index. The closure object's first field contains the index. This makes function call consistent. For closure object, the rest of the fields contain extra arguments, more stuffs needs to be pushed to stack machine before `call_indirect`.

Reply:
> We can use the extra-argument infrastructure for method calls to pass an extra reference into the function that holds all the nonlocal variables.

@jpolitz Did you mean that every compiled wasm function should have exactly one extra parameter (except potential things needed by the fancy-calling group), which is a pointer 4 bytes greater than the function value (or the same as the function value to harness recursion)? It is similar to the `self` arguments in the `apply` method of [closure class implementation](https://ucsd-cse231-w21.github.io/closures/#from-nested-functions-to-closures).
Another operation is to load nonlocal variables one by one outside, before the indirect function call, and passes all of them as separated function arguments, as said by Yaosen in the last sentence.

The first interpretation/implementation is base on the feature that nested functions/closures can be used as `callable` variables, stripping the information of its extra arguments, which makes it impossible to apply different numbers of extra arguments to different functions. Consider the following example: (modified from [nested functions lecture](https://ucsd-cse231-w21.github.io/nested-functions/))
``` python
def f(x : int) -> int:
  def g(y : int) -> int:
    return x + y

  def h(y : int) -> int:
    return 1 + y

  var_callable: Callable[[int], int] = None

  if True:  # suppose in a nontrivial example, the condition is too complex to predict at compile time
    var_callable = g  # the heap memory containing func table index of f_g
  else:
    var_callable = h  # the heap memory containing func table index of f_h, maybe followed by the reference (*if x's value is changed after this line but before calling*) of x

  # now we don't know how many/if necessary separate extra arguments we need to add to var_callable at compile time
  return var_callable(10)

f(5)
```

Reply:
Right – I don't think this can work with arbitrary numbers of extra arguments for escaping functions with nonlocals. It has to be a single heap location, because of this kind of example.